.copilot/cpp_instructions.md

```diff
@@ -0,0 +1,88 @@
+* **Thread Safety:** Pay close attention to thread safety. Shared data should be protected by appropriate synchronization primitives (e.g., `seastar::mutex`, `std::atomic`), or structured to avoid sharing entirely (e.g., using message passing or per-core data).
+* **`co_await` Semantics:** When using `co_await`, ensure you understand the execution context and potential for context switching.
+* **Avoid Blocking:** Never perform blocking I/O or long-running computations on the reactor thread. Such operations must be offloaded to dedicated thread pools.
+
+# 5. Performance and Optimization
+
+* **Zero-Overhead Abstractions:** Prefer abstractions that have zero or minimal runtime overhead.
+* **Avoid Dynamic Allocation:** Minimize dynamic memory allocations, especially in hot code paths. Prefer stack-based or pre-allocated objects.
+* **Compiler Optimizations:** Write code that is easy for the compiler to optimize. Avoid aliasing, use `constexpr` where possible, and prefer standard library algorithms that can be vectorized.
+* **Benchmarking:** Any performance-critical change should be accompanied by benchmarks to validate its impact.
+
+# 6. Testing
+
+* **Test-Driven Development (TDD)**: When modifying existing functionality, if there is a corresponding test file, follow a TDD approach. This means first creating a new test case that reproduces the bug or verifies the new feature, then making the minimum amount of changes in the implementation to pass the test.
+* **Test File Naming**: C++ unit tests often match the name of the source file they are testing (e.g., my_module_test.cc for my_module.cc).
+* **Unit Tests (C++):** When feasible, generate a native C++ unit test. For asynchronous code, use the Seastar testing macros like `SEASTAR_TEST_CASE` and `SEASTAR_THREAD_TEST_CASE`. A test case should be self-contained and test a single unit of functionality.
+    * **Example of an asynchronous Seastar test:**
+      ```cpp
+      #include <seastar/testing/test_case.hh>
+      #include <seastar/core/future.hh>
+      #include "path/to/my_header.hh"
+
+      SEASTAR_TEST_CASE(my_async_test_case) {
+          // Arrange
+          my_class_name my_instance;
+
+          // Act & Assert
+          return my_instance.my_async_method().then([](int result) {
+              BOOST_CHECK_EQUAL(result, 10);
+          });
+      }
+      ```
+* **Integration Tests (Python):** If a native C++ unit test cannot be easily created, offer a Python integration test. This test should operate at a higher level, interacting with the system as a black box (e.g., through a command-line interface or a network protocol).
+    * The Python test should be in the format currently used in the codebase, likely using `pytest` or a custom framework. The test should start the C++ executable, interact with it, and assert the correct behavior.
```

Comment:
When you write a new test, you do not "start the C++ executable", the test framework already does this for you.
Some test directories include a README.md file which an inquisitive AI (or human!) can read to understand how these tests worked, such as test/cqlpy/README.md